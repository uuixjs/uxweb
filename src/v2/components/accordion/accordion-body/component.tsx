import { Background, BorderRadiusProps, Display, Layout } from "../../layout";
import { FC, ReactNode } from "react";

import { getAriaProps } from "lib";

export interface AccordionBodyProps {
  /**
   * @example Lorem ipsum dolor sit amet, consectetur adipiscing elit
   */
  children?: ReactNode;
  /**
   * @example true
   */
  isOpen?: boolean;
  borderRadius?: BorderRadiusProps;
}

/**
 * AccordionBody is a content region whose visibility should be
 * controlled by an AccordionHeader
 */
export const AccordionBody: FC<AccordionBodyProps> = (props) => (
  <Layout
    padding={2}
    background={Background.Base}
    elevation={1}
    display={props.isOpen ? Display.Block : Display.Hide}
    borderRadius={props.borderRadius}
    {...getAriaProps(props)}
  >
    {props.children}
  </Layout>
);

AccordionBody.displayName = "AccordionBody";
