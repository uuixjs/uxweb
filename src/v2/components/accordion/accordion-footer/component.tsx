import { Button, ButtonProps, ButtonType } from "../../button/button";
import { Display, JustifyContent, Layout } from "../../layout";

import { FC } from "react";
import { getAriaProps } from "lib/ui-utils";

export interface AccordionFooterProps {
  primaryButtonProps: ButtonProps;
  secondaryButtonProps?: ButtonProps;
}

/**
 * AccordionFooter should be placed within an AccordionBody and
 * composes either one or two call-to-action buttons with consistent placement.
 */
export const AccordionFooter: FC<AccordionFooterProps> = (props) => {
  const primaryButton = (
    // Order of attributes matters; later props will override earlier props
    <Button {...props.primaryButtonProps} variant={ButtonType.Primary} />
  );

  let secondaryButton;

  if (props.secondaryButtonProps) {
    secondaryButton = (
      // Order of attributes matters; later props will override earlier props
      <Button {...props.secondaryButtonProps} variant={ButtonType.Secondary} />
    );
  }

  return (
    <Layout
      display={Display.Flex}
      justifyContent={JustifyContent.End}
      margin={{ top: 2 }}
      {...getAriaProps(props)}
    >
      {secondaryButton && (
        <Layout margin={{ left: 1 }}>{secondaryButton}</Layout>
      )}

      <Layout margin={{ left: 1 }}>{primaryButton}</Layout>
    </Layout>
  );
};

AccordionFooter.displayName = "AccordionFooter";
