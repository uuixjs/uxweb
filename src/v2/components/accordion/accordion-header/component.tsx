import { Align, Aspect, AspectRatio } from "../../aspect";
import {
  AlignItems,
  Background,
  BorderRadiusProps,
  Color,
  Display,
  FlexDirection,
  InjectLayout,
  JustifyContent,
  Layout,
  Overflow,
  Position,
  ZIndex,
} from "../../layout";
import { Component, MouseEvent } from "react";
import { CoreImage, CoreImageProps } from "../../core-image";
import { CoreText, TextType } from "../../core-text";
import {
  ExpandableIndicator,
  ExpandableIndicatorDirection,
} from "../../expandable-indicator";
import { PaddingValue, getAriaProps, styled } from "lib";
import { Title, TitleSize } from "../../title";

import { Interactable } from "../../interactable";

export enum AccordionHeaderSize {
  Default = "default",
  Narrow = "narrow",
}

export interface AccordionDefaultHeaderProps {
  index?: number;
  /**
   * @example What is thy bidding?
   */
  title: string;
  /**
   * @example Reach 10 followers
   */
  description?: string;
  isOpen?: boolean;
  onClick?: (event: MouseEvent<HTMLElement>, index?: number) => void;
  imageProps?: CoreImageProps;
  imageRatio?: AspectRatio.Aspect1x1 | AspectRatio.Aspect3x4;
  backgroundImageProps?: CoreImageProps;
  imageLabelOverlay?: string;
  label?: string;
  borderRadius?: BorderRadiusProps;
  size?: AccordionHeaderSize.Default;
}

export interface AccordionNarrowHeaderProps {
  index?: number;
  title: string;
  isOpen?: boolean;
  onClick?: (event: MouseEvent<HTMLElement>, index?: number) => void;
  borderRadius?: BorderRadiusProps;
  size: AccordionHeaderSize.Narrow;
}

const defaultHeight = "8rem";
const narrowHeight = "5rem";

const ScAccordionHeaderContents = styled(Layout)<{
  $size: AccordionHeaderSize;
}>(({ $size }) =>
  $size && $size === AccordionHeaderSize.Default
    ? `min-height: ${defaultHeight};`
    : `height: ${narrowHeight};`,
);

const ScAccordionHeaderImageContainer = styled(Layout)<{
  $ratio: AccordionDefaultHeaderProps["imageRatio"];
}>(({ $ratio }) =>
  $ratio && $ratio === AspectRatio.Aspect3x4
    ? `min-width: calc(${defaultHeight} * (3/4));`
    : `min-width: ${defaultHeight};`,
);

const ScAccordionHeaderBackgroundImage = styled(CoreImage)`
  opacity: 0.5;
  object-fit: cover;
  object-position: center;
`;

export type AccordionHeaderProps =
  | AccordionDefaultHeaderProps
  | AccordionNarrowHeaderProps;

/**
 * An AccordionHeader is an interactable region which should be used to toggle
 * the visibility of additional information.
 */
export class AccordionHeader extends Component<AccordionHeaderProps, {}> {
  public render() {
    return (
      <Layout
        className="accordion-header"
        background={Background.Base}
        elevation={1}
        position={Position.Relative}
        borderRadius={this.props.borderRadius}
        {...getAriaProps(this.props)}
      >
        {this.renderBackgroundImage()}

        <Interactable
          onClick={this.onClick}
          aria-expanded={this.props.isOpen}
          borderRadius={this.props.borderRadius}
        >
          <ScAccordionHeaderContents
            $size={
              this.props.size === AccordionHeaderSize.Default ||
              this.props.size === undefined
                ? AccordionHeaderSize.Default
                : AccordionHeaderSize.Narrow
            }
            display={Display.Flex}
            position={Position.Relative}
            borderRadius={this.props.borderRadius}
            overflow={Overflow.Hidden}
          >
            {this.renderImageAndLabel()}
            {this.renderTitles()}
            {this.renderLabel()}
            {this.renderCaret()}
          </ScAccordionHeaderContents>
        </Interactable>
      </Layout>
    );
  }

  private getYPadding = (): PaddingValue => {
    if (isAccordionHeaderNarrowProps(this.props)) {
      return 2;
    }
    return this.props.description ? 2 : 3;
  };

  private renderCaret = () => {
    return (
      <Layout
        flexGrow={0}
        flexShrink={0}
        display={Display.Flex}
        alignItems={AlignItems.Center}
        justifyContent={JustifyContent.Center}
        padding={{ y: this.getYPadding(), right: 2 }}
      >
        <ExpandableIndicator
          open={this.props.isOpen}
          openDirection={ExpandableIndicatorDirection.Up}
          closedDirection={ExpandableIndicatorDirection.Down}
        />
      </Layout>
    );
  };

  private renderTitles = () => {
    return (
      <Layout
        flexGrow={1}
        flexShrink={1}
        display={Display.Flex}
        flexDirection={FlexDirection.Column}
        alignItems={AlignItems.Start}
        justifyContent={JustifyContent.Center}
        overflow={Overflow.Hidden}
        padding={{ y: this.getYPadding(), left: 2, right: 1 }}
      >
        <Layout fullWidth>
          <Title type={TextType.H3} size={TitleSize.Small} ellipsis>
            {this.props.title}
          </Title>
          {!isAccordionHeaderNarrowProps(this.props) && this.props.description && (
            <CoreText color={Color.Alt2} ellipsis>
              {this.props.description}
            </CoreText>
          )}
        </Layout>
      </Layout>
    );
  };

  private renderBackgroundImage = () => {
    if (isAccordionHeaderNarrowProps(this.props)) {
      return;
    }

    if (!this.props.backgroundImageProps) {
      return;
    }

    return (
      <InjectLayout
        className="accordion-header__background-image"
        position={Position.Absolute}
        attachTop
        attachLeft
        fullWidth
        fullHeight
        borderRadius={this.props.borderRadius}
      >
        <ScAccordionHeaderBackgroundImage
          {...this.props.backgroundImageProps}
        />
      </InjectLayout>
    );
  };

  private renderLabel = () => {
    if (isAccordionHeaderNarrowProps(this.props)) {
      return;
    }

    if (!this.props.label) {
      return;
    }

    return (
      <Layout
        padding={{ right: 2 }}
        flexGrow={1}
        flexShrink={0}
        display={Display.Flex}
        alignItems={AlignItems.Center}
        justifyContent={JustifyContent.End}
      >
        <CoreText>{this.props.label}</CoreText>
      </Layout>
    );
  };

  private renderImageAndLabel = () => {
    if (isAccordionHeaderNarrowProps(this.props)) {
      return;
    }

    if (!this.props.imageProps && !this.props.imageLabelOverlay) {
      return;
    }

    return (
      <ScAccordionHeaderImageContainer
        $ratio={this.props.imageRatio}
        flexGrow={0}
        flexShrink={0}
        background={Background.Alt2}
        position={Position.Relative}
      >
        {this.props.imageLabelOverlay && (
          <Layout
            position={Position.Relative}
            fullHeight
            zIndex={ZIndex.Above}
            display={Display.Flex}
            alignItems={AlignItems.Center}
            justifyContent={JustifyContent.Center}
            padding={1}
            color={this.props.imageProps ? Color.Overlay : Color.Inherit}
            background={this.props.imageProps ? Background.Overlay : undefined}
          >
            <Title>{this.props.imageLabelOverlay}</Title>
          </Layout>
        )}
        {this.props.imageProps && (
          <Layout
            position={Position.Absolute}
            fullWidth
            fullHeight
            attachTop
            attachLeft
            overflow={Overflow.Hidden}
          >
            <Aspect
              ratio={this.props.imageRatio || AspectRatio.Aspect1x1}
              align={Align.Center}
            >
              <CoreImage {...this.props.imageProps} />
            </Aspect>
          </Layout>
        )}
      </ScAccordionHeaderImageContainer>
    );
  };

  private onClick = (event: MouseEvent<HTMLElement>) => {
    if (this.props.onClick !== undefined) {
      this.props.onClick(event, this.props.index);
    }
  };
}

function isAccordionHeaderNarrowProps(
  props: AccordionDefaultHeaderProps | AccordionNarrowHeaderProps,
): props is AccordionNarrowHeaderProps {
  return props.size === AccordionHeaderSize.Narrow;
}
