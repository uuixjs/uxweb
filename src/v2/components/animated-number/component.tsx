import { Component } from "react";
import CountUp from "react-countup";
import { styled } from "lib/ui-utils";

export interface AnimatedNumberProps {
  value: number;
  formatValue: (n: number) => string;
  initialValue?: number;
  duration?: number;
  children?: undefined;
  monospaced?: boolean;
}

export interface AnimatedNumberState {
  prevValue: number;
  currentValue: number;
}

export const ScAnimatedNumber = styled(CountUp)<{
  $monospaced: AnimatedNumberProps["monospaced"];
}>(({ $monospaced }) => {
  return (
    $monospaced && {
      "font-feature-settings": `"tnum"`,
    }
  );
});
export class AnimatedNumber extends Component<
  AnimatedNumberProps,
  AnimatedNumberState
> {
  public static defaultProps: Partial<AnimatedNumberProps> = {
    duration: 0.7,
    monospaced: true,
  };

  public constructor(props: AnimatedNumberProps) {
    super(props);
    this.state = {
      currentValue: props.value,
      prevValue:
        props.initialValue !== undefined ? props.initialValue : props.value,
    };
  }

  public componentDidUpdate() {
    if (this.props.value !== this.state.currentValue) {
      this.setState((s) => ({
        currentValue: this.props.value,
        prevValue: s.currentValue,
      }));
    }
  }

  public render() {
    const {
      duration,
      formatValue,
      initialValue,
      monospaced,
      value,
      ...props
    } = this.props;
    return (
      <ScAnimatedNumber
        $monospaced={monospaced}
        duration={duration}
        end={this.state.currentValue}
        start={this.state.prevValue}
        formattingFn={formatValue}
        {...props}
      />
    );
  }
}
