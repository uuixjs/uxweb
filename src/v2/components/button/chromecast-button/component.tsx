import { Aspect, AspectRatio } from "../../aspect";
import { ButtonIcon, ButtonIconSize } from "../button-icon";
import { ForwardRefRenderFunction, forwardRef } from "react";
import { keyframes, styleVariant, styled } from "lib";

import { CoreInteractivePublicProps } from "../../core-interactive";

export enum ChromecastButtonStatus {
  Available = "available",
  Connecting = "connecting",
  Connected = "connected",
}

const ScChromecastButtonIcon = styled.div`
  fill: currentColor;
`;

const ScChromecastButtonIconWifiPath = styled.path<{
  variant: "a" | "b" | "c";
  animated?: boolean;
}>`
  animation-name: ${({ animated }) =>
    animated &&
    styleVariant("variant", {
      a: undefined,
      b: animated && ScWifiBLoading,
      c: animated && ScWifiCLoading,
    })};
  animation-duration: 1.6s;
  animation-timing-function: linear;
  animation-delay: 0s;
  animation-iteration-count: infinite;
`;

const ScWifiBLoading = keyframes`
  0% {
    opacity: 0.3;
  }
  33% {
    opacity: 0.3;
  }
  34% {
    opacity: 1;
  }
  100% {
    opacity: 1;
  }
`;

const ScWifiCLoading = keyframes`
  0% {
    opacity: 0.3;
  }
  66% {
    opacity: 0.3;
  }
  67% {
    opacity: 1;
  }
  100% {
    opacity: 1;
  }
`;

export interface ChromecastButtonProps extends CoreInteractivePublicProps {
  /**
   * Passes an `aria-label` to the component to provide context to assistive
   * devices. All standard HTML aria attributes are now supported on Core UI
   * components â€” see https://www.w3.org/TR/html-aria/."
   */
  "aria-label": string;
  status?: ChromecastButtonStatus | "available" | "connecting" | "connected"; // TODO: https://jira.xarth.tv/browse/COREUI-3554
  overlay?: boolean;
  size?: ButtonIconSize;
}

const ChromecastButton: ForwardRefRenderFunction<
  HTMLElement,
  ChromecastButtonProps
> = ({ status, ...buttonProps }, ref) => {
  const animated = status === ChromecastButtonStatus.Connecting;
  return (
    <ButtonIcon
      {...buttonProps}
      ref={ref}
      icon={
        <ScChromecastButtonIcon className="tw-chromecast-button__icon">
          <Aspect ratio={AspectRatio.Aspect1x1}>
            <svg
              width="100%"
              height="100%"
              version="1.1"
              viewBox={`0 0 20 20`}
              x="0px"
              y="0px"
            >
              <ScChromecastButtonIconWifiPath
                variant="a"
                d="M3 17a1 1 0 0 1-1-1v-1a2 2 0 0 1 2 2H3z"
              />
              <ScChromecastButtonIconWifiPath
                variant="b"
                animated={animated}
                d="M2 12.5a.5.5 0 0 1 .5-.5A4.5 4.5 0 0 1 7 16.5a.5.5 0 0 1-1 0A3.5 3.5 0 0 0 2.5 13a.5.5 0 0 1-.5-.5z"
              />
              <ScChromecastButtonIconWifiPath
                variant="c"
                animated={animated}
                d="M2 9.5a.5.5 0 0 1 .5-.5 7.5 7.5 0 0 1 7.5 7.5.5.5 0 0 1-1 0A6.5 6.5 0 0 0 2.5 10a.5.5 0 0 1-.5-.5z"
              />

              {status === ChromecastButtonStatus.Connected && (
                <path
                  className="tw-chromecast-button__screen"
                  d="M11.168 13H14V7H6v.832A10.037 10.037 0 0 1 11.168 13z"
                />
              )}

              <path
                className="tw-chromecast-button__frame"
                d="M2 5a2 2 0 0 1 2-2h12a2 2 0 0 1 2 2v10a2 2 0 0 1-2 2h-3.5a1 1 0 1 1 0-2H16V5H4v1.5a1 1 0 0 1-2 0V5z"
              />
            </svg>
          </Aspect>
        </ScChromecastButtonIcon>
      }
    />
  );
};

ChromecastButton.displayName = "ChromecastButton";
const ComponentWithRef = forwardRef(ChromecastButton);
export { ComponentWithRef as ChromecastButton };
