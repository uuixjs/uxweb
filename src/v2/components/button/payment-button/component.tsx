import {
  CoreButton,
  CoreButtonPublicProps,
  CoreButtonType,
} from "../core-button";
import { ForwardRefRenderFunction, ReactNode, forwardRef } from "react";
import {
  css,
  focusVisible,
  hoverCss,
  staticTokenRule,
  styleVariant,
  styled,
  themeTokenRule,
} from "lib/ui-utils";

export interface PaymentButtonProps extends CoreButtonPublicProps {
  children?: ReactNode;
  variant:
    | PaymentButtonType
    | "xsolla"
    | "amazon-pay"
    | "paypal-blue"
    | "paypal-yellow"; // TODO: https://jira.xarth.tv/browse/COREUI-3554
}

export enum PaymentButtonType {
  Xsolla = "xsolla",
  AmazonPay = "amazon-pay",
  PaypalBlue = "paypal-blue",
  PaypalYellow = "paypal-yellow",
}

const hoverActiveStyles = css`
  &::before {
    background-color: ${themeTokenRule(
      "color-background-interactable-alpha-hover",
    )};
  }
  &:active {
    &::before {
      background-color: ${themeTokenRule(
        "color-background-interactable-alpha-active",
      )};
    }
  }
`;
export const ScPaymentButton = styled(CoreButton)<{
  $variant:
    | PaymentButtonType
    | "xsolla"
    | "amazon-pay"
    | "paypal-blue"
    | "paypal-yellow"; // TODO: https://jira.xarth.tv/browse/COREUI-3554;
}>`
  &::before {
    content: "";
    position: absolute;
    left: 0;
    top: 0;
    right: 0;
    bottom: 0;
  }

  background-color: ${styleVariant("$variant", {
    [PaymentButtonType.AmazonPay]: staticTokenRule("color-amazon"),
    [PaymentButtonType.PaypalBlue]: staticTokenRule("color-paypal-blue"),
    [PaymentButtonType.PaypalYellow]: staticTokenRule("color-paypal-yellow"),
    [PaymentButtonType.Xsolla]: "",
  })};

  color: ${styleVariant("$variant", {
    [PaymentButtonType.AmazonPay]: staticTokenRule("color-black"),
    [PaymentButtonType.PaypalBlue]: themeTokenRule("color-text-overlay"),
    [PaymentButtonType.PaypalYellow]: staticTokenRule("color-black"),
    [PaymentButtonType.Xsolla]: "",
  })};

  ${(props) =>
    !props.disabled &&
    css`
      ${hoverCss`${hoverActiveStyles}`}
      ${focusVisible`${hoverActiveStyles}`}
    `}
`;

const ScPaymentButtonLayout = styled.div`
  padding: ${staticTokenRule("button-padding-x")};
  position: relative;
  display: flex;
  align-items: center;
`;

const PaymentButton: ForwardRefRenderFunction<
  HTMLElement,
  PaymentButtonProps
> = ({ children, variant, ...buttonProps }, ref) => {
  if (variant === PaymentButtonType.Xsolla) {
    return (
      <CoreButton
        {...buttonProps}
        ref={ref}
        className="tw-payment-button"
        variant={CoreButtonType.Secondary}
      >
        <ScPaymentButtonLayout>{children}</ScPaymentButtonLayout>
      </CoreButton>
    );
  }

  return (
    <ScPaymentButton
      {...buttonProps}
      ref={ref}
      className="tw-payment-button"
      $variant={variant}
    >
      <ScPaymentButtonLayout>{children}</ScPaymentButtonLayout>
    </ScPaymentButton>
  );
};

PaymentButton.displayName = "PaymentButton";
const ComponentWithRef = forwardRef(PaymentButton);
export { ComponentWithRef as PaymentButton };
