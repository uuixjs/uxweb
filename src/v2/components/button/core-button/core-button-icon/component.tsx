import { CoreButtonSize } from "../component";
import { FC } from "react";
import { Icon } from "../../../icon";
import { SVGAsset } from "../../../svg";
import { TEST_AUTOMATION_SELECTORS } from "../../../../tests/selectors";
import { styled } from "lib/ui-utils";

export interface CoreButtonIconProps {
  icon: SVGAsset;
  size?: CoreButtonSize | "default" | "small" | "large"; // TODO: https://jira.xarth.tv/browse/COREUI-3554
}

const SizeMap: Record<CoreButtonSize, string> = {
  [CoreButtonSize.Small]: "1.6rem",
  [CoreButtonSize.Default]: "2rem",
  [CoreButtonSize.Large]: "2.4rem",
};

const ScCoreButtonIcon = styled.div<{ size: string }>`
  display: inline-flex;
  align-items: center;
  width: ${(props) => props.size};
`;

/**
 * Buttons communicate that an action will happen when clicked. Best used as
 * primary actions inside of another component such as a Modals, Cards, and/or
 * forms. **Buttons should never be used for generic links.**
 */
export const CoreButtonIcon: FC<CoreButtonIconProps> = ({
  icon,
  size = CoreButtonSize.Default,
  ...props
}) => {
  return (
    <ScCoreButtonIcon
      className="tw-core-button-icon"
      size={SizeMap[size]}
      {...props}
    >
      <Icon
        fill
        asset={icon}
        data-a-selector={TEST_AUTOMATION_SELECTORS.CoreButtonIcon}
      />
    </ScCoreButtonIcon>
  );
};

CoreButtonIcon.displayName = "CoreButtonIcon";
