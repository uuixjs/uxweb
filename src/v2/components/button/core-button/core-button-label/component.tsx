import { AlignItems, Display, JustifyContent, Layout } from "../../../layout";
import {
  CoreButtonDropdown,
  CoreButtonDropdownType,
} from "../core-button-dropdown";
import { FC, ReactNode } from "react";
import { staticTokenRule, styled } from "lib";

import { CoreButtonIcon } from "../core-button-icon";
import { CoreButtonSize } from "..";
import { SVGAsset } from "../../../svg";
import { TEST_AUTOMATION_SELECTORS } from "../../../../tests/selectors";

export interface CoreButtonLabelProps {
  labelAlign?: JustifyContent;
  children?: ReactNode;
  fullWidth?: boolean;
  icon?: SVGAsset;
  size?: CoreButtonSize | "default" | "small" | "large"; // TODO: https://jira.xarth.tv/browse/COREUI-3554
  dropdown?: boolean | CoreButtonDropdownType;
}

interface ScCoreButtonLabelProps {
  hasIcon?: boolean;
  isDropdown?: boolean;
  isFullWidth?: boolean;
}

const ScCoreButtonLabel = styled.div<ScCoreButtonLabelProps>`
  display: flex;
  align-items: center;
  flex-grow: ${(props) => (props.isDropdown || props.isFullWidth ? 1 : 0)};
  padding-top: 0;
  padding-bottom: 0;
  padding-left: ${(props) =>
    props.hasIcon
      ? `calc(${staticTokenRule("button-padding-x")(props)} - 0.2rem)`
      : staticTokenRule("button-padding-x")(props)};
  padding-right: ${(props) =>
    props.isDropdown
      ? `calc(${staticTokenRule("button-padding-x")(props)} - 0.2rem)`
      : staticTokenRule("button-padding-x")(props)};
`;

export const CoreButtonLabel: FC<CoreButtonLabelProps> = ({
  children,
  dropdown,
  fullWidth,
  icon,
  labelAlign,
  size,
  ...props
}) => {
  return (
    <ScCoreButtonLabel
      hasIcon={!!icon}
      isFullWidth={fullWidth}
      isDropdown={!!dropdown}
      {...props}
    >
      {icon && (
        <Layout
          margin={icon && (children || dropdown) ? { right: 0.5 } : 0}
          display={Display.Flex}
          alignItems={AlignItems.Center}
        >
          <CoreButtonIcon icon={icon} size={size} />
        </Layout>
      )}
      {children && (
        <Layout
          flexGrow={dropdown || fullWidth ? 1 : 0}
          data-a-target={TEST_AUTOMATION_SELECTORS.CoreButtonLabelText}
          display={Display.Flex}
          alignItems={AlignItems.Center}
          justifyContent={labelAlign ? labelAlign : JustifyContent.Start}
        >
          {children}
        </Layout>
      )}
      {dropdown && (
        <Layout
          margin={{ left: 0.5 }}
          display={Display.Flex}
          alignItems={AlignItems.Center}
        >
          <CoreButtonDropdown
            size={size}
            type={
              dropdown === true ? CoreButtonDropdownType.ArrowDown : dropdown
            }
          />
        </Layout>
      )}
    </ScCoreButtonLabel>
  );
};
