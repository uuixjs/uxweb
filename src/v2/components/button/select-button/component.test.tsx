import { shallow } from "enzyme";
import { setupShallowTest } from "../../../tests/helpers";
import { SVGAsset } from "../../svg";
import { CoreButtonSize } from "../core-button";
import { SelectButtonComponent, SelectButtonProps } from "./component";

function getOptionalProps(): SelectButtonProps {
  return {
    autoFocus: true,
    children: "Test Button",
    disabled: true,
    download: "Download all the things",
    fullWidth: true,
    icon: SVGAsset.Gear,
    linkTo: "/",
    onClick: jest.fn(),
    overlay: true,
    size: CoreButtonSize.Large,
    tabIndex: 0,
    targetBlank: true,
    title: "Test title",
  };
}

const setupShallow = setupShallowTest(SelectButtonComponent);
describe("SelectButton", () => {
  it("renders SelectButton component with required props", () => {
    const { wrapper } = setupShallow();
    expect(wrapper).toMatchSnapshot();
  });

  it("renders SelectButton component with all props", () => {
    const { wrapper } = setupShallow(getOptionalProps());
    expect(wrapper).toMatchSnapshot();
  });

  it("passes aria props to the element", () => {
    const wrapper = shallow(<SelectButtonComponent aria-label="Some Action" />);
    const element = wrapper.find("[aria-label]").first();
    expect(element.prop("aria-label")).toBe("Some Action");
  });
});
