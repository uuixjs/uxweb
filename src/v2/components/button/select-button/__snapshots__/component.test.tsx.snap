// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectButton renders SelectButton component with all props 1`] = `
<Styled(CoreButton)
  autoFocus={true}
  className="tw-select-button tw-select-button--disabled tw-select-button--overlay"
  disabled={true}
  download="Download all the things"
  fullWidth={true}
  icon="Gear"
  linkTo="/"
  onClick={[MockFunction]}
  overlay={true}
  size="large"
  tabIndex={0}
  targetBlank={true}
  title="Test title"
>
  <CoreButtonLabel
    dropdown={1}
    fullWidth={true}
    icon="Gear"
    size="large"
  >
    Test Button
  </CoreButtonLabel>
</Styled(CoreButton)>
`;

exports[`SelectButton renders SelectButton component with required props 1`] = `
<Styled(CoreButton)
  className="tw-select-button"
>
  <CoreButtonLabel
    dropdown={1}
  />
</Styled(CoreButton)>
`;
